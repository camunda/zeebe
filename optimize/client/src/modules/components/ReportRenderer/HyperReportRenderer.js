/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import update from 'immutability-helper';

import {formatters, getReportResult} from 'services';

import CombinedReportRenderer from './CombinedReportRenderer';
import ProcessReportRenderer from './ProcessReportRenderer';

const {formatReportResult} = formatters;

export default function HyperReportRenderer({report, ...rest}) {
  const result = getReportResult(report);

  if (!result.data.length || result.instanceCount === 0) {
    const data = result.data.map((entry) => ({...entry, value: 0}));
    const emptyReport = update(report, {result: {measures: {$set: [{data}]}}});
    return <ProcessReportRenderer {...rest} report={emptyReport} />;
  }

  const firstEntryResult = result.data[0].value;
  const newResultData = {};

  formatResult(report.data, firstEntryResult).forEach(({key, label}) => {
    newResultData[key] = {
      combined: false,
      id: key,
      name: label,
      reportType: 'process',
      data: report.data,
      result: {
        ...result,
        type: 'map',
        measures: result.measures.map((measure) => ({
          ...measure,
          type: 'map',
          data: measure.data.map((datapoint) => ({
            ...datapoint,
            value: datapoint.value.find((data) => data.key === key)?.value,
          })),
        })),
      },
    };
  });

  const convertedReport = {
    ...report,
    combined: true,
    data: {
      configuration: report.data.configuration,
      reports: firstEntryResult.map(({key}) => ({id: key})),
      visualization: report.data.visualization,
    },
    result: {
      ...result,
      type: 'hyperMap',
      data: newResultData,
    },
  };

  return <CombinedReportRenderer {...rest} report={convertedReport} />;
}

function formatResult(data, result) {
  const {
    distributedBy,
    configuration: {distributeByDateVariableUnit},
  } = data;

  const distributedByDateVar =
    distributedBy.type === 'variable' && distributedBy.value.type === 'Date';
  const distributedByDate = ['startDate', 'endDate'].includes(distributedBy.type);

  if (distributedByDate || distributedByDateVar) {
    return formatReportResult(
      {
        ...data,
        groupBy: distributedBy,
        configuration: {
          ...data.configuration,
          groupByDateVariableUnit: distributeByDateVariableUnit,
        },
      },
      result
    );
  }

  return result;
}
