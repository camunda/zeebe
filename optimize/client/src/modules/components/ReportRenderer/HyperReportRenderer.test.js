/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import HyperReportRenderer from './HyperReportRenderer';

it('should convert a hypermap to a combined report', () => {
  const node = shallow(
    <HyperReportRenderer
      report={{
        combined: false,
        reportType: 'process',
        data: {
          processDefinitionKey: 'aKey',
          processDefinitionVersion: '1',
          view: {
            properties: ['duration'],
            entity: 'userTask',
          },
          groupBy: {
            type: 'assignee',
          },
          visualization: 'bar',
          distributedBy: {type: 'userTask', value: null},
          configuration: {},
        },
        result: {
          measures: [
            {
              property: 'duration',
              data: [
                {
                  key: 'Anne',
                  label: 'Anne',
                  value: [
                    {key: 'taskId1', label: 'Usertask 1', value: 8},
                    {key: 'taskId2', label: 'Usertask 2', value: 1},
                    {key: 'taskId3', label: 'Usertask 3', value: 65},
                  ],
                },
                {
                  key: 'Bernd',
                  label: 'Bernd',
                  value: [
                    {key: 'taskId1', label: 'Usertask 1', value: 3},
                    {key: 'taskId2', label: 'Usertask 2', value: 17},
                    {key: 'taskId3', label: 'Usertask 3', value: 22},
                  ],
                },
                {
                  key: 'Chris',
                  label: 'Chris',
                  value: [
                    {key: 'taskId1', label: 'Usertask 1', value: 1},
                    {key: 'taskId2', label: 'Usertask 2', value: 0},
                    {key: 'taskId3', label: 'Usertask 3', value: 73},
                  ],
                },
              ],
            },
          ],
          instanceCount: 1234,
          type: 'hyperMap',
        },
      }}
    />
  );

  expect(node).toMatchSnapshot();
});

it('should render single process report if report result is empty', () => {
  const node = shallow(
    <HyperReportRenderer
      report={{
        result: {
          measures: [{data: []}],
          instanceCount: 0,
          type: 'hyperMap',
        },
      }}
    />
  );

  expect(node).toMatchSnapshot();
});

it('should format single reports names for hyper reports distributed by date variable', () => {
  const node = shallow(
    <HyperReportRenderer
      report={{
        combined: false,
        reportType: 'process',
        data: {
          processDefinitionKey: 'aKey',
          processDefinitionVersion: '1',
          view: {
            properties: ['frequency'],
            entity: 'processInstance',
          },
          groupBy: {
            type: 'startDate',
          },
          visualization: 'bar',
          distributedBy: {
            type: 'variable',
            value: {type: 'Date'},
          },
          configuration: {
            distributeByDateVariableUnit: 'month',
          },
        },
        result: {
          measures: [
            {
              property: 'frequency',
              data: [
                {
                  key: '2020-09-01T00:00:00.000+0200',
                  label: '2020-09-01T00:00:00.000+0200',
                  value: [
                    {
                      key: '1969-12-07T14:52:00Z',
                      value: 1.0,
                      label: '1969-12-07T14:52:00Z',
                    },
                    {
                      key: '2020-03-08T14:52:00Z',
                      value: 1.0,
                      label: '2020-03-08T14:52:00Z',
                    },
                  ],
                },
              ],
            },
          ],
          instanceCount: 1234,
          type: 'hyperMap',
        },
      }}
    />
  );

  const names = Object.values(node.prop('report').result.data).map(({name}) => name);

  expect(names).toEqual(['Dec 1969', 'Mar 2020']);
});
