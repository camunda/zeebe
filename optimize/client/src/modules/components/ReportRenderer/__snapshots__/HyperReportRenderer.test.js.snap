// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should convert a hypermap to a combined report 1`] = `
<CombinedReportRendererErrorHandler
  report={
    Object {
      "combined": true,
      "data": Object {
        "configuration": Object {},
        "reports": Array [
          Object {
            "id": "taskId1",
          },
          Object {
            "id": "taskId2",
          },
          Object {
            "id": "taskId3",
          },
        ],
        "visualization": "bar",
      },
      "reportType": "process",
      "result": Object {
        "data": Object {
          "taskId1": Object {
            "combined": false,
            "data": Object {
              "configuration": Object {},
              "distributedBy": Object {
                "type": "userTask",
                "value": null,
              },
              "groupBy": Object {
                "type": "assignee",
              },
              "processDefinitionKey": "aKey",
              "processDefinitionVersion": "1",
              "view": Object {
                "entity": "userTask",
                "properties": Array [
                  "duration",
                ],
              },
              "visualization": "bar",
            },
            "id": "taskId1",
            "name": "Usertask 1",
            "reportType": "process",
            "result": Object {
              "data": Array [
                Object {
                  "key": "Anne",
                  "label": "Anne",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 8,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 1,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 65,
                    },
                  ],
                },
                Object {
                  "key": "Bernd",
                  "label": "Bernd",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 3,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 17,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 22,
                    },
                  ],
                },
                Object {
                  "key": "Chris",
                  "label": "Chris",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 1,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 0,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 73,
                    },
                  ],
                },
              ],
              "instanceCount": 1234,
              "measures": Array [
                Object {
                  "data": Array [
                    Object {
                      "key": "Anne",
                      "label": "Anne",
                      "value": 8,
                    },
                    Object {
                      "key": "Bernd",
                      "label": "Bernd",
                      "value": 3,
                    },
                    Object {
                      "key": "Chris",
                      "label": "Chris",
                      "value": 1,
                    },
                  ],
                  "property": "duration",
                  "type": "map",
                },
              ],
              "type": "map",
            },
          },
          "taskId2": Object {
            "combined": false,
            "data": Object {
              "configuration": Object {},
              "distributedBy": Object {
                "type": "userTask",
                "value": null,
              },
              "groupBy": Object {
                "type": "assignee",
              },
              "processDefinitionKey": "aKey",
              "processDefinitionVersion": "1",
              "view": Object {
                "entity": "userTask",
                "properties": Array [
                  "duration",
                ],
              },
              "visualization": "bar",
            },
            "id": "taskId2",
            "name": "Usertask 2",
            "reportType": "process",
            "result": Object {
              "data": Array [
                Object {
                  "key": "Anne",
                  "label": "Anne",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 8,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 1,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 65,
                    },
                  ],
                },
                Object {
                  "key": "Bernd",
                  "label": "Bernd",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 3,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 17,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 22,
                    },
                  ],
                },
                Object {
                  "key": "Chris",
                  "label": "Chris",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 1,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 0,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 73,
                    },
                  ],
                },
              ],
              "instanceCount": 1234,
              "measures": Array [
                Object {
                  "data": Array [
                    Object {
                      "key": "Anne",
                      "label": "Anne",
                      "value": 1,
                    },
                    Object {
                      "key": "Bernd",
                      "label": "Bernd",
                      "value": 17,
                    },
                    Object {
                      "key": "Chris",
                      "label": "Chris",
                      "value": 0,
                    },
                  ],
                  "property": "duration",
                  "type": "map",
                },
              ],
              "type": "map",
            },
          },
          "taskId3": Object {
            "combined": false,
            "data": Object {
              "configuration": Object {},
              "distributedBy": Object {
                "type": "userTask",
                "value": null,
              },
              "groupBy": Object {
                "type": "assignee",
              },
              "processDefinitionKey": "aKey",
              "processDefinitionVersion": "1",
              "view": Object {
                "entity": "userTask",
                "properties": Array [
                  "duration",
                ],
              },
              "visualization": "bar",
            },
            "id": "taskId3",
            "name": "Usertask 3",
            "reportType": "process",
            "result": Object {
              "data": Array [
                Object {
                  "key": "Anne",
                  "label": "Anne",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 8,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 1,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 65,
                    },
                  ],
                },
                Object {
                  "key": "Bernd",
                  "label": "Bernd",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 3,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 17,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 22,
                    },
                  ],
                },
                Object {
                  "key": "Chris",
                  "label": "Chris",
                  "value": Array [
                    Object {
                      "key": "taskId1",
                      "label": "Usertask 1",
                      "value": 1,
                    },
                    Object {
                      "key": "taskId2",
                      "label": "Usertask 2",
                      "value": 0,
                    },
                    Object {
                      "key": "taskId3",
                      "label": "Usertask 3",
                      "value": 73,
                    },
                  ],
                },
              ],
              "instanceCount": 1234,
              "measures": Array [
                Object {
                  "data": Array [
                    Object {
                      "key": "Anne",
                      "label": "Anne",
                      "value": 65,
                    },
                    Object {
                      "key": "Bernd",
                      "label": "Bernd",
                      "value": 22,
                    },
                    Object {
                      "key": "Chris",
                      "label": "Chris",
                      "value": 73,
                    },
                  ],
                  "property": "duration",
                  "type": "map",
                },
              ],
              "type": "map",
            },
          },
        },
        "instanceCount": 1234,
        "measures": Array [
          Object {
            "data": Array [
              Object {
                "key": "Anne",
                "label": "Anne",
                "value": Array [
                  Object {
                    "key": "taskId1",
                    "label": "Usertask 1",
                    "value": 8,
                  },
                  Object {
                    "key": "taskId2",
                    "label": "Usertask 2",
                    "value": 1,
                  },
                  Object {
                    "key": "taskId3",
                    "label": "Usertask 3",
                    "value": 65,
                  },
                ],
              },
              Object {
                "key": "Bernd",
                "label": "Bernd",
                "value": Array [
                  Object {
                    "key": "taskId1",
                    "label": "Usertask 1",
                    "value": 3,
                  },
                  Object {
                    "key": "taskId2",
                    "label": "Usertask 2",
                    "value": 17,
                  },
                  Object {
                    "key": "taskId3",
                    "label": "Usertask 3",
                    "value": 22,
                  },
                ],
              },
              Object {
                "key": "Chris",
                "label": "Chris",
                "value": Array [
                  Object {
                    "key": "taskId1",
                    "label": "Usertask 1",
                    "value": 1,
                  },
                  Object {
                    "key": "taskId2",
                    "label": "Usertask 2",
                    "value": 0,
                  },
                  Object {
                    "key": "taskId3",
                    "label": "Usertask 3",
                    "value": 73,
                  },
                ],
              },
            ],
            "property": "duration",
          },
        ],
        "type": "hyperMap",
      },
    }
  }
/>
`;

exports[`should render single process report if report result is empty 1`] = `
<ProcessReportRenderer
  report={
    Object {
      "result": Object {
        "instanceCount": 0,
        "measures": Array [
          Object {
            "data": Array [],
          },
        ],
        "type": "hyperMap",
      },
    }
  }
/>
`;
